conjurapi/authn_test.go

```diff
@@ -1,92 +1,114 @@
-		So(err, ShouldBeNil)
+type testCase struct {
+	name             string
+	roleId           string
+	login            string
+	readResponseBody bool
+}
 
-		conjur.LoadPolicy(
-			PolicyModePut,
-			"root",
-			strings.NewReader(policy),
-		)
+func TestClient_RotateAPIKey(t *testing.T) {
+	testCases := []testCase{
+		{
+			name:             "Rotate the API key of a foreign user role of kind user",
+			roleId:           "cucumber:user:alice",
+			login:            "alice",
+			readResponseBody: false,
+		},
+		{
+			name:             "Rotate the API key of a foreign role of non-user kind",
+			roleId:           "cucumber:host:bob",
+			login:            "host/bob",
+			readResponseBody: false,
+		},
+		{
+			name:             "Rotate the API key of a foreign role and read the data stream",
+			roleId:           "cucumber:user:alice",
+			login:            "alice",
+			readResponseBody: true,
+		},
+	}
 
-		Convey("Rotate the API key of a foreign user role of kind user", func() {
-			aliceAPIKey, err := conjur.RotateAPIKey("cucumber:user:alice")
+	t.Run("V5", func(t *testing.T) {
 
-			_, err = conjur.Authenticate(authn.LoginPair{"alice", string(aliceAPIKey)})
-			So(err, ShouldBeNil)
-		})
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				// SETUP
+				config := &Config{}
+				config.mergeEnv()
 
-		Convey("Rotate the API key of a foreign role of non-user kind", func() {
-			bobAPIKey, err := conjur.RotateAPIKey("cucumber:host:bob")
+				apiKey := os.Getenv("CONJUR_AUTHN_API_KEY")
+				login := os.Getenv("CONJUR_AUTHN_LOGIN")
 
-			_, err = conjur.Authenticate(authn.LoginPair{"host/bob", string(bobAPIKey)})
-			So(err, ShouldBeNil)
-		})
+				policy := `
+- !user alice
+- !host bob
+`
 
-		Convey("Rotate the API key of a foreign role and read the data stream", func() {
-			rotateResponse, err := conjur.RotateAPIKeyReader("cucumber:user:alice")
+				conjur, err := NewClientFromKey(*config, authn.LoginPair{Login: login, APIKey: apiKey})
+				assert.NoError(t, err)
 
-			So(err, ShouldBeNil)
-			aliceAPIKey, err := ReadResponseBody(rotateResponse)
-			So(err, ShouldBeNil)
+				conjur.LoadPolicy(
+					PolicyModePut,
+					"root",
+					strings.NewReader(policy),
+				)
 
-			_, err = conjur.Authenticate(authn.LoginPair{"alice", string(aliceAPIKey)})
-			So(err, ShouldBeNil)
-		})
+				// EXERCISE
+				runAssertions(t, tc, config, login, apiKey)
+			})
+		}
 	})
 
 	if os.Getenv("TEST_VERSION") != "oss" {
-		Convey("V4", t, func() {
-
-			config := &Config{
-				ApplianceURL: os.Getenv("CONJUR_V4_APPLIANCE_URL"),
-				SSLCert:      os.Getenv("CONJUR_V4_SSL_CERTIFICATE"),
-				Account:      os.Getenv("CONJUR_V4_ACCOUNT"),
-				V4:           true,
-			}
+		t.Run("V4", func(t *testing.T) {
 
-			login := os.Getenv("CONJUR_V4_AUTHN_LOGIN")
-			apiKey := os.Getenv("CONJUR_V4_AUTHN_API_KEY")
+			for _, tc := range testCases {
+				t.Run(tc.name, func(t *testing.T) {
+					// SETUP
+					config := &Config{
```

Comment:
Lines 78-86 (for the V4 test case) also do not depend upon the test case config, so they can be moved outside of the for loop into a `SETUP` section that is common to all V4 test cases.

Or move these to a separate function.

Also, if you also include these first 2 lines from `runAssertions()` in this common `SETUP`:
```
	conjur, err := NewClientFromKey(*config, authn.LoginPair{Login: login, APIKey: apiKey})
	assert.NoError(t, err)
```
then you can eliminate 3 arguments to the `runAssertions()` function.